### DISCUSSION TITLE
Support for non-HTML output formats?

### DISCUSSION THREAD

**Discussion by shah** [lumeland/lume#113] [Q&A] [answered]
I was curious if Lume supports non-HTML output formats such as:

* iCal (if we want to generate an iCal dynamically and serve that up)
* RSS (for obvious reasons)
* JSON (for example to serve up a JSON file through generated content, not static)
* MD (in case the final markdown should be served to the client and transformed in the browser)

By way of reference, here the [custom output formats Hugo supports](https://gohugo.io/templates/output-formats/) (we are using iCal and a few others now).

**Comment by oscarotero** [comment, marked as answer]
Yes, you can output any format. There are two ways:

- Add a subextension to the file. For example, `styles.css.njk` is a nunjucks file that output the file `styles.css`.
- Or include the `url` variable in the frontmatter or named module. For example:
```ts
export const url = "feed.rss";
```

In base-blog you have examples of both methods:
- Nunjucks template that exports a xml file with the RSS. [Link](https://github.com/lumeland/base-blog/blob/master/sitemap.xml.njk)
- Javascript template that exports a Json Feed. [Link](https://github.com/lumeland/base-blog/blob/master/feed.tmpl.js)

**Reply by shah** [reply]
Beautiful, thanks!

**Comment by shah** [comment]
In case it's helpful, for those using TypeScript this is necessary if creating a JSON output format: `/api/file.json.tmpl.ts` template will create `/api/file.json` (I did not have a layout and just emitted output of `JSON.stringify(data)`. 

The feature is fantastic and much easier than in other static site generators. Now, we can create complete APIs statically.
